server/Main.hs

```diff
@@ -53,31 +53,36 @@ instance A.ToJSON Error
 
 server :: [P.ExternsFile] -> P.Env -> P.Environment -> Int -> IO ()
 server externs initNamesEnv initEnv port = do
+  let toCompilerErrors :: NE.NonEmpty CST.ParserError -> Either Error a
+      toCompilerErrors = Left . CompilerErrors . P.toJSONErrors False P.Error . CST.toMultipleErrors "<file>"
+
   let compile :: Text -> IO (Either Error ([P.JSONError], JS))
       compile input
         | T.length input > 20000 = return (Left (OtherError "Please limit your input to 20000 characters"))
         | otherwise = do
-          case CST.parseModuleFromFile "<file>" input >>= CST.resFull of
+          case CST.parseModuleFromFile "<file>" input of
             Left parseError ->
-              return . Left . CompilerErrors . P.toJSONErrors False P.Error $ CST.toMultipleErrors "<file>" parseError
-            Right m | P.getModuleName m == P.ModuleName "Main" -> do
-              (resultMay, ws) <- runLogger' . runExceptT . flip runReaderT P.defaultOptions $ do
-                ((P.Module ss coms moduleName elaborated exps, env), nextVar) <- P.runSupplyT 0 $ do
-                  desugared <- P.desugar initNamesEnv externs [P.importPrim m] >>= \case
-                    [d] -> pure d
-                    _ -> error "desugaring did not produce one module"
-                  P.runCheck' (P.emptyCheckState initEnv) $ P.typeCheckModule desugared
-                regrouped <- P.createBindingGroups moduleName . P.collapseBindingGroups $ elaborated
-                let mod' = P.Module ss coms moduleName regrouped exps
-                    corefn = CF.moduleToCoreFn env mod'
-                    [renamed] = P.renameInModules [corefn]
-                unless (null . CF.moduleForeign $ renamed) . throwError . P.errorMessage $ P.MissingFFIModule moduleName
-                P.evalSupplyT nextVar $ P.prettyPrintJS <$> J.moduleToJs renamed Nothing
-              case resultMay of
-                Left errs -> (return . Left . CompilerErrors . P.toJSONErrors False P.Error) errs
-                Right js -> (return . Right) (P.toJSONErrors False P.Error ws, js)
-            Right _ ->
-              (return . Left . OtherError) "The name of the main module should be Main."
+              return $ toCompilerErrors parseError
+            Right partialResult -> case CST.resFull partialResult of
+              (_, Left parseError) ->
+                return $ toCompilerErrors parseError
+              (warnings, Right m) | P.getModuleName m == P.ModuleName "Main" -> do
```

Comment:
It might help to call the warnings variable `parserWarnings`? Just because there's quite a lot that goes on in between it being bound here and it being used at the end of this expression.